src/lib.rs

```diff
@@ -1,241 +1,188 @@
-            connect_internal_flash: rom_data::connect_internal_flash::ptr(),
-            flash_exit_xip: rom_data::flash_exit_xip::ptr(),
-            flash_range_erase: if erase {
-                Some(rom_data::flash_range_erase::ptr())
-            } else {
-                None
-            },
-            flash_range_program: if write {
-                Some(rom_data::flash_range_program::ptr())
-            } else {
-                None
-            },
-            flash_flush_cache: rom_data::flash_flush_cache::ptr(),
-            flash_enter_cmd_xip: rom_data::flash_enter_cmd_xip::ptr(),
-            phantom: PhantomData,
-        }
+    pub struct FlashIAP {
+        fn_ptrs: FlashFunctionPointers,
+        boot2: [u32; 256 / 4],
     }
 
-    #[allow(unused)]
-    /// # Safety
-    ///
-    /// `boot2` must contain a valid 2nd stage boot loader which can be called to re-initialize XIP mode
-    unsafe fn flash_function_pointers_with_boot2(
-        erase: bool,
-        write: bool,
-        boot2: &[u32; 64],
-    ) -> FlashFunctionPointers {
-        let boot2_fn_ptr = (boot2 as *const u32 as *const u8).offset(1);
-        let boot2_fn: unsafe extern "C" fn() -> () = core::mem::transmute(boot2_fn_ptr);
-        FlashFunctionPointers {
-            connect_internal_flash: rom_data::connect_internal_flash::ptr(),
-            flash_exit_xip: rom_data::flash_exit_xip::ptr(),
-            flash_range_erase: if erase {
-                Some(rom_data::flash_range_erase::ptr())
-            } else {
-                None
-            },
-            flash_range_program: if write {
-                Some(rom_data::flash_range_program::ptr())
-            } else {
-                None
-            },
-            flash_flush_cache: rom_data::flash_flush_cache::ptr(),
-            flash_enter_cmd_xip: boot2_fn,
-            phantom: PhantomData,
+    impl FlashIAP {
+        pub fn new(use_boot2: bool) -> Self {
+            let mut me = Self {
+                fn_ptrs: FlashFunctionPointers {
+                    connect_internal_flash: rom_data::connect_internal_flash::ptr(),
+                    flash_exit_xip: rom_data::flash_exit_xip::ptr(),
+                    flash_range_erase: rom_data::flash_range_erase::ptr(),
+                    flash_range_program: rom_data::flash_range_program::ptr(),
+                    flash_flush_cache: rom_data::flash_flush_cache::ptr(),
+                    flash_enter_cmd_xip: rom_data::flash_enter_cmd_xip::ptr(),
+                },
+                boot2: [0u32; 256 / 4],
+            };
+            if use_boot2 {
+                unsafe {
+                    rom_data::memcpy44(&mut me.boot2 as *mut _, 0x10000000 as *const _, 256);
+                    let boot2_fn_ptr = (&me.boot2 as *const u32 as *const u8).offset(1);
+                    let boot2_fn: unsafe extern "C" fn() -> () = core::mem::transmute(boot2_fn_ptr);
+                    me.fn_ptrs.flash_enter_cmd_xip = boot2_fn;
```

Comment:
I don't think this is correct: This essentially creates a pointer into `Self`, but `Self` can move at any time. I guess that this already breaks if you disable optimizations, but even with optimizations, it's not guaranteed that the struct is created in-place and not copied to the caller.

Reply:
hm that's true, I did'nt think about that.
This resolution (lines 43-45) can be moved to `write_flash_inner` so that it's computed right before jumping to ram.